### ISSUE TITLE
Bug: [no-invalid-void-type] Can't use `void` in return type union for overloaded function

### ISSUE THREAD

**Issue by haines** [typescript-eslint/typescript-eslint#5752]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.8.4&sourceType=module&code=GYVwdgxgLglg9mABMAFASgFyIG5xgEwG4AoUSWBZFADywGcoAnGMAc00QebZLOniSpqAfnpMW7LLgKIAPp3FtEAb2KJEMYIhqIAhAF59icPgCmwFqfxoVa9YkamoIRkmol1AX2LfS4fpSs6FgACoxwALYwdKYAPNL4AHy8-hRIQbQK3JKIYZHRcVwSycQAhnQAnpDIqQKIGaJZEhx5UTHxePhyTWyJtuqa2tR6hsZgZhZgVjaq9g5OLm4eiN6eQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uYDcBDRgBNaPDpWFFS6KImjQO0SODABfECqA&tsconfig=N4XyA

### Repro Code

```typescript
function f(): void;
function f(x: string): string;
function f(x?: string): void | string {
  if (x !== undefined) {
    return x;
  }
}

function g(): Promise<void>;
function g(x: string): Promise<string>;
async function g(x?: string): Promise<void | string> {
  if (x !== undefined) {
    return x;
  }
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-invalid-void-type": "error"
  }
}
```

### tsconfig

```jsonc
{}
```

### Expected Result

No errors.

### Actual Result

void is only valid as a return type or generic type variable. 3:25 - 3:29
void is only valid as a return type or generic type variable. 11:39 - 11:43

### Additional Info

It's possible to work around the linter in the first example by changing the implementation return type from `void | ...` to `undefined | ...`, although this may require introducing an explicit `return undefined` statement to the function body if the compiler option `noImplicitReturns` is set.

In the second example, changing `Promise<void | ...>` to `Promise<undefined | ...>` is not possible because it introduces a compiler error

>This overload signature is not compatible with its implementation signature. (2394)

### Versions

| package | version |
| -- | -- |
| `@typescript-eslint/eslint-plugin` | `5.39.0` |
| `@typescript-eslint/parser` | `5.39.0` |
| `TypeScript` | `4.8.4` |
| `ESLint` | `8.15.0` |
| `node` | `web` |

**Comment by JoshuaKGoldberg** [comment]
Interesting. Normally the `void | string` return type would be a direct "no" from this rule... but it's also the only way to support overloads. Accepting PRs for this case!

**Comment by lsdsjy** [comment]
> Normally the `void | string` return type would be a direct "no" from this rule...

Why is that? I think the following should be allowed:
```ts
function g(x?: string): void | string {
  if (x !== undefined && x.length > 10) {
    return x;
  }
}
```
It's identical to the function overloading case: either you keep `void` in return type, or you change `void` to `undefined` and manually return `undefined` in the end of the function.

**Comment by bradzacher** [comment]
> Why is that? I think the following should be allowed

Because by unioning `void` you're conflating types and concepts - which is what the rule is there to prevent.

It's explained in the rule's docs
https://typescript-eslint.io/rules/no-invalid-void-type/ 

If that's not a code style you want to enforce - then feel free to disable the rule as it's not for you!

**Comment by bradzacher** [comment]
> But if it's valid in the overloading case, it should also be valid without overloading.

I disagree.

If you're declaring a function with an overload you're explicitly declaring a case in which a function return can be ignored (eg some combination of parameters).

Whilst an overload does ultimately result in a function implementation that is declared with a void union (if you're explicitly typing the return), it is is vastly different in style and intetion to just declaring any old function with a void union.

Put another way with an overload you're saying "in this specific case the function will return nothing, and in this other case it returns something" whereas in the non-overload case you're saying "this function might return nothing or something" - vastly different meanings!

**Comment by pfeileon** [comment]
What about this case:
```
const foo = new Subject<void | string>();
```

**Comment by Michael-Ziluck** [comment]
What @pfeileon is describing is also my use case. I want a `Subject` to be able to take a `string` or no input to the `next` method call but this lint rule causes an error. For now I'm just suppressing the rule but it would be super nice to get that fixed as part of this issue.

**Comment by guillaumebrunerie** [comment]
I think void unions should always be allowed in return types (even without overloads), as far as I know it is the only way to say that a function should either return a specific type or return nothing, which I think is a valid use case.
```ts
type ReturnNothingOrNumber = () => void | number;

// Should work
const f1: ReturnNothingOrNumber = () => {};

// Should work
const f2: ReturnNothingOrNumber = () => {
    return 42;
};

// Should *not* work
const f3: ReturnNothingOrNumber = () => {
    return "42";
};
```

With `() => undefined | number`, the first function does not typecheck (type `void` is not assignable to type `undefined | number`).

I’ve stumbled upon this issue working on a wrapper function around immer. The `recipe` argument to the `produce` function in immer is a function that should either not return anything (most common case), or return something of a specific type.

**Comment by bradzacher** [comment]
@guillaumebrunerie by design of the `void` type - returning `string` is completely valid. That's intentional and by design, for better or worse.

In a "receiving" position - `() => void` doesn't mean "a function that returns nothing" it means "a function whose return type I will not be using".

By that definition if you want "a function that returns nothing" - `undefined` is literally the most correct type to use.

**Comment by guillaumebrunerie** [comment]
@bradzacher Well, that's what I would have expected too, but the fact is that
- the function `f3` returning a string does not typecheck (as desired), even though one might think that returning a string should be valid when `void` is part of the return type
- changing the return type to `undefined | number` makes the function `f1` not typecheck, even though not returning anything should be the same as returning `undefined`.

Maybe it's a bug in Typescript (or two different bugs), but using a void union seems to be the only way to get the behavior I want, currently.
If there is another way to get Typescript to accept `f1` and `f2` but not `f3`, I'd like to know about it.

**Comment by bradzacher** [comment]
@guillaumebrunerie taking a step back - let me rephrase.

If you think of `void` in this location as "I don't care about the return type" (as mentioned before) then the type `() => void | number` means "a function that returns a number or I don't care about the return type". Which is kind of non-sensical, right?

And on the other side you would probably have a check like `if (returnValue != undefined)` - which shows you do, in fact, care about the return type!

Additionally this type isn't as safe as you might think because it's still quite easy to pass a string function into your type - eg
```ts
const f4 = () => "";
const f5: () => void = f4;
const f6: ReturnNothingOrNumber = f5;
```

Using `void` just isn't what you want to be correct and safe!

**Comment by guillaumebrunerie** [comment]
@bradzacher I fully agree with you, it is non-sensical and I wish that `undefined | number` would work, but it doesn't. Which is a bit strange because I thought they fixed it in Typescript 5.1.
I guess in this case the issue is that functions without return statement are inferred as returning `void`, so we lose the information that it does in fact return `undefined`.

**Comment by jun-sheaf** [comment]
> If you think of `void` in this location as "I don't care about the return type" (as mentioned before) then the type `() => void | number` means "a function that returns a number or I don't care about the return type". Which is kind of non-sensical, right?

I think as a consumer of a function, it's more accurate to say when a function returns `void`, then the return type is UB rather than "I don't care" or "return type I will not be using". So `number | void` really means if TypeScript deduces a `number`, then you can be certain about the return type. If TypeScript deduces `void`, then it's UB. This then implies "return type should not be used" (or "I don't care").

**Comment by JoshuaKGoldberg** [comment]
Clarifying question: what is "UB"?

**Comment by jun-sheaf** [comment]
> Clarifying question: what is "UB"?

Undefined Behavior

**Comment by bradzacher** [comment]
> The distinction is important since it implies as a function implementor, you can extend the return type later without breaking changes; e.g. number | void to number | string | void.

But this is incorrect and would lead you to breaking users code in a minor.

Whilst TS doesn't treat it as exactly `undefined` - TS does still treat `void` as a "nullish" type. Which means if your consumer uses a `!= null` it will refine away the `void`.

```ts
declare function foo(): number | void;
const x = foo();
//    ^?
if (x != null) {
  x
//^? Number
  x.toFixed();
} else {
  x
//^? Void
}
```
[playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtVJxwAoBKALnlWQFsAjEGeAH3gDcctgBuAKDDwDOGeAA94AXgLESvAPSz4i+AD0A-DyyJ4RMQEJJVCBBLwA3j0Uie8tRdEA6DDgBiWESGCleAX3ggIAgjmltaytt48QA)

So the model of explicitly saying "this is undefined behaviour" is incongruous with how typescript treats the type. If you add `string` to that signature you would change the behaviour of the code and cause the user's code to crash at runtime (and fail typechecking).

In terms of breaking changes - using `| void` is no better than using `| undefined`.

**Comment by jun-sheaf** [comment]
> TS does still treat void as a "nullish" type. Which means if your consumer uses a != null it will refine away the void.

By saying this you just ignored my statement about UB. Whether or not TS treats it as "nullish" is irrelevant as it's UB.

A correct usage of `number | void` is the following:

```ts
declare function foo(): number | void;
const x = foo();
//    ^?
if (typeof x === 'number') {
  x
//^? Number
  x.toFixed();
} else {
  x
//^? Void
}
```

If the return type switches to `number | string | void`, then the above statement still works, so it's not a breaking change. `| undefined` isn't great because someone can write:

```ts
declare function foo(): number | undefined;
const x = foo();
//    ^?
if (typeof x !== 'undefined') {
  x
//^? number
  x.toFixed();
} else {
  x
//^? undefined
}
```

This breaks if the return type switches to `number | string | undefined`.

**Comment by bradzacher** [comment]
> Btw, your code at runtime has an unhandled case because x := 0 means x is 0 in the else branch

That is incorrect. `x != null` does not treat falsey values as false.
It returns false if and only if `x` is `null | undefined` - i.e. it is equivalent to `x !== null && x !== undefined`.

----

I would like to be clear here:

This issue has nothing at all to do with the general case of `| void` in return types. That is a topic for a separate issue.
This issue is specifically for allowing `| void` in the return type of a function that has an overloaded declaration that returns a bare `void`.

We are accepting a PR to support the latter.
The former is not up for debate in this issue and is off-topic.
If the former is something that interests you - please file a separate issue.

**Comment by jun-sheaf** [comment]
I've reopened https://github.com/typescript-eslint/typescript-eslint/issues/8755 then. It's a generalization of this issue, so I'm not sure what you want to do with this issue.

Also, `!= null` is the same as falsy value. You can test this on the DevTools console.

**Comment by Josh-Cena** [comment]
While this is also related to https://github.com/typescript-eslint/typescript-eslint/issues/8113, I think this can progress separately. We just need to ignore the implementation signature if overload signatures exist.

**Comment by jedwards1211** [comment]
Similar use case: constraining the return type of a method in a base class

```ts
class MQTTConnection {
  _authorizePublish(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    packet: IPublishPacket
  ):
    | (boolean | null | undefined | void)
    | Promise<boolean | null | undefined | void> {
    // no-op unless overridden
  }
  ...
}
```

**Comment by dandv** [comment]
I'm confused by whether a void union should be allowed for overloaded functions, which is what @bradzacher is [saying in this comment](https://github.com/typescript-eslint/typescript-eslint/issues/5752#issuecomment-1315188452).

But the code and playground in [the first comment](https://github.com/typescript-eslint/typescript-eslint/issues/5752#issue-1399653412) are exactly that, two overloads (with and without Promise), and the rule was and is still triggered. So it seems void unions are *not* allowed for function overloads either? Or am I missing something?

My use case is equivalent to this:

```
function g(): Promise<void>;
function g(x: string): Promise<string>;
async function g(x?: string): Promise<void | string> {
  if (x !== undefined) {
    return x;
  }
}
```

The rule triggers: `void is not valid as a constituent in a union type 3:39 - 3:43`. How should one define the signature of `g`?

**Comment by jedwards1211** [comment]
I'm tempted to just turn this rule off for myself, I don't feel like I'm getting a lot of benefit from it.  Does anyone have examples of it saving them from major bugs they wouldn't have caught otherwise?
